presto-hive/src/main/java/com/facebook/presto/hive/S3SelectRecordCursorProvider.java

```diff
@@ -0,0 +1,102 @@
+import java.util.Set;
+
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_FILESYSTEM_ERROR;
+import static com.facebook.presto.hive.HiveUtil.getDeserializerClassName;
+import static java.util.Objects.requireNonNull;
+
+public class S3SelectRecordCursorProvider
+        implements HiveRecordCursorProvider
+{
+    public static final String S3_SELECT_PUSHDOWN_ENABLED = "s3.select-pushdown.enabled";
+
+    private static final Set<String> CSV_SERDES = ImmutableSet.of("org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe");
+    private final HdfsEnvironment hdfsEnvironment;
+    private final HiveClientConfig clientConfig;
+    private final PrestoS3ClientFactory s3ClientFactory;
+
+    @Inject
+    public S3SelectRecordCursorProvider(
+            HdfsEnvironment hdfsEnvironment,
+            HiveClientConfig clientConfig,
+            PrestoS3ClientFactory s3ClientFactory)
+    {
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.clientConfig = requireNonNull(clientConfig, "clientConfig is null");
+        this.s3ClientFactory = requireNonNull(s3ClientFactory, "s3ClientFactory is null");
+    }
+
+    @Override
+    public Optional<RecordCursor> createRecordCursor(
+            Configuration configuration,
+            ConnectorSession session,
+            Path path,
+            long start,
+            long length,
+            long fileSize,
+            Properties schema,
+            List<HiveColumnHandle> columns,
+            TupleDomain<HiveColumnHandle> effectivePredicate,
+            DateTimeZone hiveStorageTimeZone,
+            TypeManager typeManager)
+    {
+        boolean isS3SelectPushdownEnabled = configuration.getBoolean(S3_SELECT_PUSHDOWN_ENABLED, false);
+        if (!isS3SelectPushdownEnabled) {
+            return Optional.empty();
+        }
+
+        try {
+            this.hdfsEnvironment.getFileSystem(session.getUser(), path, configuration);
+        }
+        catch (IOException e) {
+            throw new PrestoException(HIVE_FILESYSTEM_ERROR, "Failed getting filesystem for path: " + path, e);
+        }
+
+        String serdeName = getDeserializerClassName(schema);
+        if (CSV_SERDES.contains(serdeName)) {
+            IonSqlQueryBuilder queryBuilder = new IonSqlQueryBuilder(typeManager);
+            String ionSqlQuery = buildQuery(queryBuilder, columns, effectivePredicate);
+            S3SelectLineRecordReader recordReader = new S3SelectCsvRecordReader(configuration, clientConfig, path, start, length, schema, ionSqlQuery, s3ClientFactory);
+            return Optional.of(new S3SelectRecordCursor(configuration, path, recordReader, length, schema, columns, hiveStorageTimeZone, typeManager));
+        }
+
+        // unsupported serdes
+        return Optional.empty();
+    }
+
+    private String buildQuery(IonSqlQueryBuilder queryBuilder, List<HiveColumnHandle> columns, TupleDomain<HiveColumnHandle> effectivePredicate)
```

Comment:
This method doesn't seem useful since it's just delegating to another method. Let's remove it.

Reply:
refactored